atlasexec/atlas_test.go

```diff
@@ -51,7 +59,180 @@ func Test_MigrateApply(t *testing.T) {
+					Text: `Dropping table "t2"`,
+					Code: "DS102",
+				}},
+			}},
+			Error: "destructive changes detected",
+		}
+		require.EqualValues(t, expectedReport, got.Files[0])
+	})
+	t.Run("lint with manually parsing output", func(t *testing.T) {
+		c, err := atlasexec.NewClient(".", "atlas")
+		require.NoError(t, err)
+		var buf bytes.Buffer
+		err = c.MigrateLintError(context.Background(), &atlasexec.MigrateLintParams{
+			DevURL: "sqlite://file?mode=memory",
+			DirURL: "file://testdata/migrations",
+			Latest: 1,
+			Writer: &buf,
+		})
+		require.NoError(t, err)
+		var raw json.RawMessage
+		require.NoError(t, json.NewDecoder(&buf).Decode(&raw))
+	})
+}
+
+type graphQLQuery struct {
+	Query     string          `json:"query"`
+	Variables json.RawMessage `json:"variables"`
+}
+
+type Dir struct {
+	Name    string `json:"name"`
+	Content string `json:"content"`
+}
+type dirsQueryResponse struct {
+	Data struct {
+		Dirs []Dir `json:"dirs"`
+	} `json:"data"`
+}
+
+func TestMigrateLintWithLogin(t *testing.T) {
+	token := "123456789"
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		require.Equal(t, "Bearer "+token, r.Header.Get("Authorization"))
+		query := graphQLQuery{}
+		err := json.NewDecoder(r.Body).Decode(&query)
+		require.NoError(t, err)
+		switch {
+		case strings.Contains(query.Query, "mutation reportMigrationLint"):
+			fmt.Fprintf(w, `{ "data": { "reportMigrationLint": { "url": "https://migration-lint-report-url" } } }`)
+		}
+	}))
+	t.Cleanup(srv.Close)
+	atlasConfigURL := generateHCL(t, srv.URL, token)
+	t.Run("Web and Writer params produces an error", func(t *testing.T) {
+		c, err := atlasexec.NewClient(".", "atlas")
+		require.NoError(t, err)
+		params := &atlasexec.MigrateLintParams{
+			ConfigURL: atlasConfigURL,
+			DevURL:    "sqlite://file?mode=memory",
+			DirURL:    "file://testdata/migrations",
+			Latest:    1,
+		}
+		paramsWeb := *params
+		paramsWeb.Web = true
+		got, err := c.MigrateLint(context.Background(), &paramsWeb)
+		require.ErrorContains(t, err, "custom Writer or Web reporting are not supported")
+		require.Nil(t, got)
+		paramsWriter := *params
+		paramsWriter.Writer = &bytes.Buffer{}
+		got, err = c.MigrateLint(context.Background(), &paramsWriter)
+		require.ErrorContains(t, err, "custom Writer or Web reporting are not supported")
+		require.Nil(t, got)
+	})
+	t.Run("lint parse web output", func(t *testing.T) {
+		c, err := atlasexec.NewClient(".", "atlas")
+		require.NoError(t, err)
+		var buf bytes.Buffer
+		err = c.MigrateLintError(context.Background(), &atlasexec.MigrateLintParams{
+			DevURL: "sqlite://file?mode=memory",
+			DirURL: "file://testdata/migrations",
+
+			ConfigURL: atlasConfigURL,
+			Latest:    1,
+			Writer:    &buf,
+			Web:       true,
+		})
+		require.NoError(t, err)
+		require.Equal(t, strings.TrimSpace(string(buf.Bytes())), "https://migration-lint-report-url")
+	})
+}
+
+func generateHCL(t *testing.T, url, token string) string {
```

Comment:
No need for this extra function as it's used only once. You can also avoid using templates and just pass the URL as a variable to Atlas.

Reply:
I disagree here.

Having a function that is used just once is completely fine as long as it serves it's purpose; Making the code maintainable and readable. This function has a clear responsibility and when reading the code that is using it, you can immediately understand the purpose of the author. 

On the other hand, if we inline the function, the reader now must understand the implementation details of this function in-order to understand the calling code

That being said - if there is a way to avoid the template, I'll do it. I'm not aware of how I can pass the URL and token to atlas without an HCL file. This is a pattern that I saw in many of our tests and I'd love to change that if you can refer me to an example.

Reply:
> Having a function that is used just once is completely fine as long as it serves it's purpose; Making the code maintainable and readable. This function has a clear responsibility and when reading the code that is using it, you can immediately understand the purpose of the author.

The problem I have with this is the amount of line jumping you have to do in order to understand what is happening. Especially in review (which is 90% the cause for reading code) it is quite cumbersome. If you wanna skip the implementation to read / review the outer function, every IDE and even GitHub has the possibility to collapse parts. I am with Ariel here.

Reply:
I think differently about almost all of these points

- code read during review is mostly a one-time thing, it's going to remain in the codebase for much longer
- it's going to have more impact on readability when we try to make changes to the code
- understanding what's going on does not require jumping around unless you are specifically looking for the implementation details. Most of the time you are skimming through the code and you want to understand what's the intention, not how it's implemented

I think Martin Fowler summarized it really well - https://martinfowler.com/bliki/FunctionLength.html
It's also a big topic in the [Clean Code book ](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882)

I'm going to inline this function, and will try to align to our practices, but I love the discussion that's going on here :)

Reply:
https://github.com/ariga/atlas-go-sdk/pull/18/commits/5a9c64293d35a960b8ed3b26e5abe662541eae43

@a8m @masseelch please re-review